### ISSUE TITLE
_poserror is reserved

### ISSUE THREAD

**Issue by pmor13** [cc65/cc65#1796]
This C code does not violate any syntax rule or constraint:
```
#define _poserror 0
#include <stdio.h>
```
Hence, it is required to be successfully compiled. However, cc65 fails to successfully compile it. This is because cc65's `stdio.h` reserves `_poserror`. However, the end user may use `_poserror` as a macro name.

**Comment by pmor13** [comment]
Ah, cc65 is non-conforming:
```
src/cc65/compile.c:399:    /* DefineNumericMacro ("__STDC__", 1);      <- not now */
```
Hence, this issue is "Not a problem".

**Comment by mrdudz** [comment]
We are still aiming for standard compliance, so it IS a bug :)

**Comment by groessler** [comment]
Can you point me to where it is stated that `_poserror` is a reserved name?

**Comment by polluks2** [comment]
reserved by underscore

**Comment by spiro-trikaliotis** [comment]
According to n1256, 6.10.8 no. 4, only underscore and an upper-case letter or a second underscore fall under the category "reserved".

**Comment by groessler** [comment]
Also 7.1.3 says "All identifiers that begin with an underscore are always reserved for use as identifiers
with file scope in both the ordinary and tag name spaces."
Don't know what's meant by "ordinary and tag name spaces".

I always considered identifiers starting with "_" to be implementation defined, so don't use them in my own code...

**Comment by spiro-trikaliotis** [comment]
You're right.

Tag name space is the name is the tags of a struct, union. That is, in struct ABC {....} , the ABC ist the tag name.

I also considered everything beginning with an underscore as reserved for the implementation, but I could not find it in the spec. This, I thought I was over-cautious.

**Comment by groessler** [comment]
So,

- if `_poserror` is reserved the OT makes a mistake and the ticket is invalid.
- if `_poserror` is not reserved it's a bug in CC65

We could rename `_poserror` to `_POSERROR` to be on the safe side. Looks ugly though.

**Comment by groessler** [comment]
or `__poserror`, looks nicer

**Comment by mrdudz** [comment]
Whatever you work out - please consider creating a test for it :)

**Comment by groessler** [comment]
Puh What should be tested there?

**Comment by mrdudz** [comment]
If invalid names are rejected as they should be? And that the valid ones that are rejected now do then work?

**Comment by pmor13** [comment]
@groessler
1. The issue title (usually) summarizes the problem. The problem is that `_poserror` is reserved. Hence, the title says "_poserror is reserved".
2. Per C the `_poserror` is not reserved. Hence, reserving of `_poserror` is a bug in CC65.

**Comment by groessler** [comment]
I'm still not quite sure/convinced that it really isn't reserved by the standard.

**Comment by pmor13** [comment]
Identifier `_poserror` being used as macro name does not fall into 7.1.3 Reserved identifiers (C11).

**Comment by kugelfuhr** [comment]
15 comments and even an already merged PR because no one reads the standard. Listen to the experts :hear_no_evil:

Let me do some reading for you.

ISO/IEC 9899:1990 (E) (this is the C89 standard document) says in 7.1.3 _"Reserved identifiers"_ (page 97) among other things:

> All identifiers that begin with an underscore are always reserved as identifiers with file scope in both the ordinary identifier and tag name spaces.

To clarify what "file scope" is, we have to look into 6.1.2.1 _"Scopes of identifiers"_ on page 20:

> If the declarator or type specifier that declares the identifier appears outside of any block or list of parameters, the identifier has file scope, which terminates at the end of the translation unit.

So these two sentences do clearly mean that `_poserror` has file scope and is therefore reserved by the C standard as an identifier. But there's more in 7.1.3. It reads later:

> If the program declares or defines an identifier with the same name as an identifier reserved in that context (other than as allowed by 7.1.7), the behavior is undefined.

Footnote 91 is attached to this sentence. It reads:

> Since macro names are replaced whenever found, independent of scope and name space, macro names matching one of the reserved identifier names must not be defined if an associated header file, if any, is used.

This does __exactly__ cover the case in this issue. A macro is defined with the name of a reserved identifier. This is not allowed.

Just for completeness let me look up the topic in newer standards:

ISO/IEC 9899:1999 (E) (the C99 standard) says in 7.1.3p1 (page 166):

> All identifiers that begin with an underscore are always reserved for use as identifiers with file scope in both the ordinary and tag name spaces.

So this is the same as in the C89 standard. Instead of the footnote it reads in 7.1.3p2:

> If the program declares or defines an identifier in a context in which it is reserved (other than as allowed by 7.1.4), or defines a reserved identifier as a macro name, the behavior is undefined.

So the new standard is mostly identical, but explicitly states undefined behavior in the given case.

In the C11 standard, ISO/IEC 9899:2011 (E), wording and chapter numbers are identical to that of the C99 standard.

To summarize:

  * `_poserror` is a reserved identifier and defining a macro with the name of a reserved identifier triggers undefined behavior.

  * The bug report is therefore invalid and the issue can be closed.

  * PR #1805 is useless and can be reverted.

  * The developers should read the C standard.

**Comment by pmor13** [comment]
@kugelfuhr Wrong.

C11, 6.2.1 Scopes of identifiers::
> **Macro names and macro parameters are not considered further here**, because prior to the semantic phase of
program translation any occurrences of macro names in the source file are replaced by the preprocessing token sequences that constitute their macro definitions.

**Comment by groessler** [comment]
Where is the next available language lawyer?

**Comment by pmor13** [comment]
P.J. Plauger, Standard C Library Name Space Control, JCLT, 1(4), page 251:
> Remember that the user can write macros that begin with an underscore, followed by a lowercase letter.

**Comment by mrdudz** [comment]
> https://github.com/cc65/cc65/pull/1805 is useless and can be reverted.

You are right, it was late, i will revert it.

**Comment by pmor13** [comment]
Another example:
```
#include <errno.h>	
int f(void)
{
    static struct { int x; } _errno;
    return errno;
}
```
leads to:
```
<source>:5: Error: Converting to 'int' from 'struct $anon-struct-0001'
```
while it does not violate any syntax rule or constraint.

**Comment by mrdudz** [comment]
To reiterate: First of all i should say i am only interested in solving the original problem here, not in whether the standard actually dictates us to do so. IE the standard might allow the "_symbol" type of name to be reserved - or not (thats how i interpret it at least), and it may require the "__symbol" type of name to be reserved. In that case renaming _symbol to __symbol - while not dictated by the standard - would remove the problem, and still satisfy the standard. To say it in other words, it is undisputed that "__symbol" is reserved. We can evade the whole dispute about whether "_symbol" is reserved or not by renaming those symbols to "__symbol". (Please correct me if am wrong). In that case i'd vote for renaming those symbols. I'd also propose searching for other cases like this and rename as many as possible in one PR. Plus there ideally is a test program that checks for at least those cases that were found+fixed.

**Comment by kugelfuhr** [comment]
> Another example:

This one is very different from your original report and in this case you're correct. The reason is that the macro `errno` is defined by the implementation, and declaration of `_errno` is in block scope, not file scope.

I would suggest to create a separate issue for it.

> We can evade the whole dispute about whether "_symbol" is reserved or not by renaming those symbols to "__symbol". (Please correct me if am wrong). In that case i'd vote for renaming those symbols. I'd also propose searching for other cases like this and rename as many as possible in one PR. 

Supported.

**Comment by groessler** [comment]
So re-integrate #1805?

Tell me what else to change. I have a nice script to replace strings in a directory tree....

**Comment by mrdudz** [comment]
I cant reopen that PR, so if you want to put some time into fixing this, please make a new one (sorry).

This might be a start (some of these are likely struct members, i have not checked them). Please don't forget to check first if a symbol with two underscores doesnt already exist :)

```
./atari.h:223:extern void __fastcall__ _setcolor     (unsigned char color_reg, unsigned char hue, unsigned char luminance);
./atari.h:224:extern void __fastcall__ _setcolor_low (unsigned char color_reg, unsigned char color_value);
./atari.h:225:extern unsigned char __fastcall__ _getcolor (unsigned char color_reg);
./atari.h:232:extern int  __fastcall__ _graphics (unsigned char mode); /* mode value same as in BASIC */
./atari.h:233:extern void __fastcall__ _scroll (signed char numlines);
./atari.h:244:extern void _save_vecs(void);                /* save system vectors */
./atari.h:245:extern void _rest_vecs(void);                /* restore system vectors */
./atari.h:247:extern unsigned char _is_cmdline_dos(void);  /* does DOS support command lines */
./atari.h:254:extern unsigned char _dos_type;         /* the DOS flavour */
./atari.h:545:#define _textcolor(color)        COLOR_WHITE
./stdlib.h:95:void __fastcall__ _heapadd (void* mem, size_t size);
./stdlib.h:98:size_t __fastcall__ _heapblocksize (const void* block);
./stdlib.h:101:size_t _heapmemavail (void);
./stdlib.h:104:size_t _heapmaxavail (void);
./stdlib.h:112:void _randomize (void);         /* Non-standard */
./stdlib.h:133:void __fastcall__ _swap (void* p, void* q, size_t size);
./assert.h:45:extern void __fastcall__ _afailed (const char*, unsigned);
./assert.h:46:#  define assert(expr)  ((expr)? (void)0 : _afailed(__FILE__, __LINE__))
./cc65.h:88:int __fastcall__ _sin (unsigned x);
./cc65.h:93:int __fastcall__ _cos (unsigned x);
./_atari5200os.h:57:    unsigned char _free_1[0xEF];     // = $11-$FF   User space
./conio.h:203:#ifdef _textcolor
./conio.h:204:#  define textcolor(color)      _textcolor(color)
./conio.h:206:#ifdef _bgcolor
./conio.h:207:#  define bgcolor(color)        _bgcolor(color)
./conio.h:209:#ifdef _bordercolor
./conio.h:210:#  define bordercolor(color)    _bordercolor(color)
./conio.h:212:#ifdef _cpeekcolor
./conio.h:213:#  define cpeekcolor()          _cpeekcolor()
./conio.h:215:#ifdef _cpeekrevers
./conio.h:216:#  define cpeekrevers()         _cpeekrevers()
./time.h:92:clock_t _clocks_per_sec (void);
./time.h:93:#  define CLOCKS_PER_SEC        _clocks_per_sec()
./lynx.h:207:#define _MIKEY_TIMERS (*(struct _mikey_all_timers *) 0xFD00)  // mikey_timers[8]
./lynx.h:208:#define _HBL_TIMER (*(struct _mikey_timer *) 0xFD00)          // timer0 (HBL)
./lynx.h:209:#define _VBL_TIMER (*(struct _mikey_timer *) 0xFD08)          // timer2 (VBL)
./lynx.h:210:#define _UART_TIMER (*(struct _mikey_timer *) 0xFD14)         // timer4 (UART)
./cbm.h:91:extern char _filetype;          /* Defaults to 's' */
./setjmp.h:45:int __fastcall__ _setjmp (jmp_buf buf);
./setjmp.h:46:#define setjmp  _setjmp         /* ISO insists on a macro */
./_mikey.h:31:typedef struct _mikey_timer {
./_mikey.h:36:} _mikey_timer;
./_mikey.h:38:typedef struct _mikey_all_timers {
./_mikey.h:39:  struct _mikey_timer timer[8];
./_mikey.h:40:} _mikey_all_timers;
./_mikey.h:43:typedef struct _mikey_audio {
./_mikey.h:52:} _mikey_audio;
./_mikey.h:56:  struct _mikey_timer timer0;       // 0xFD00
./_mikey.h:57:  struct _mikey_timer timer1;       // 0xFD04
./_mikey.h:58:  struct _mikey_timer timer2;       // 0xFD08
./_mikey.h:59:  struct _mikey_timer timer3;       // 0xFD0C
./_mikey.h:60:  struct _mikey_timer timer4;       // 0xFD10
./_mikey.h:61:  struct _mikey_timer timer5;       // 0xFD14
./_mikey.h:62:  struct _mikey_timer timer6;       // 0xFD18
./_mikey.h:63:  struct _mikey_timer timer7;       // 0xFD1C
./_mikey.h:64:  struct _mikey_audio channel_a;    // 0xFD20
./_mikey.h:65:  struct _mikey_audio channel_b;    // 0xFD28
./_mikey.h:66:  struct _mikey_audio channel_c;    // 0xFD30
./_mikey.h:67:  struct _mikey_audio channel_d;    // 0xFD38
./sym1.h:131:struct _display {
./sym1.h:140:#define DISPLAY (*(struct _display*)0xA640)
./string.h:77:void* __fastcall__ _bzero (void* ptr, size_t n);
./string.h:95:const char* __fastcall__ _stroserror (unsigned char errcode);
./apple2_filetype.h:318:extern unsigned char _filetype;  /* Default: PRODOS_T_BIN */
./apple2_filetype.h:319:extern unsigned int _auxtype;    /* Default: 0            */
./apple2.h:124:extern unsigned char _dos_type;
./apple2.h:167:} _datetime;
./apple2.h:207:#define _textcolor(color)       COLOR_WHITE
./apple2.h:208:#define _bgcolor(color)         COLOR_BLACK
./apple2.h:209:#define _bordercolor(color)     COLOR_BLACK
./apple2.h:210:#define _cpeekcolor()           COLOR_WHITE
./apple2.h:211:#define _cpeekrevers()          0
./osic1p.h:43:#define _textcolor(color)       COLOR_WHITE
./osic1p.h:44:#define _bgcolor(color)         COLOR_BLACK
./osic1p.h:45:#define _bordercolor(color)     COLOR_BLACK
./stdio.h:150:void __fastcall__ _poserror (const char* msg);          /* cc65 */
./errno.h:48:extern unsigned char _oserror;
./errno.h:50:extern int _errno;
./errno.h:53:#define errno   _errno
./errno.h:86:int __fastcall__ _osmaperrno (unsigned char oserror);
./errno.h:91:unsigned char __fastcall__ _seterrno (unsigned char code);
./errno.h:94:int __fastcall__ _directerrno (unsigned char code);
./errno.h:99:int __fastcall__ _mappederrno (unsigned char code);
./_heap.h:40:extern unsigned*          _heaporg;     /* Bottom of heap */
./_heap.h:41:extern unsigned*          _heapptr;     /* Current top */
./_heap.h:42:extern unsigned*          _heapend;     /* Upper limit */
./_heap.h:43:extern struct freeblock*  _heapfirst;   /* First free block in list */
./_heap.h:44:extern struct freeblock*  _heaplast;    /* Last free block in list */
./telestrat.h:129:#define _bordercolor(color)     COLOR_BLACK
./signal.h:48:void __fastcall__ _sig_ign (int);
./signal.h:49:void __fastcall__ _sig_dfl (int);
./signal.h:52:#define SIG_DFL         _sig_dfl
./signal.h:53:#define SIG_IGN         _sig_ign
./6502.h:95:void __fastcall__ _sys (struct regs* r);
./_atarios.h:292:    unsigned char   _free_1[0xD4-0x7F-1];   // USER SPACE
./_atarios.h:365:    unsigned char _spare_1;                 // = $0233          No OS use.
./_atarios.h:407:    unsigned char _reserved_1[29];          // = $024E-$026A    RESERVED
./_atarios.h:446:    unsigned char _reserved_2[5];           // = $028B-$028F    RESERVED
./_atarios.h:493:    unsigned char _spare_2[23];             // = $02C9-$02DF    No OS use.
./_atarios.h:532:    unsigned char _spare_3;                 // = $02E9          No OS use.
./_atarios.h:550:    unsigned char _spare_4[5];              // = $02F5-$02F9    No OS use.
./_atarios.h:577:    unsigned char _spare_5;                 // = $0313          unused
./_atarios.h:602:    unsigned char _spare_6[151];            // = $03E8-$047F    unused
./_atarios.h:623:    unsigned char _spare_7[126];            // = $0500-$057D    reserved for FP package / unused
./_atarios.h:643:    unsigned char   _internal_1[0xBA-0x91-1];  // INTERNAL DATA
./_atarios.h:647:    unsigned char   _internal_2[0xC3-0xBB-1];   // INTERNAL DATA
./_atarios.h:651:    unsigned char   _internal_3[0xC9-0xC3-1];   // INTERNAL DATA
./_atarios.h:656:    unsigned char   _internal_4[0xD4-0xCA-1];   // INTERNAL DATA
./atari5200.h:94:#define _bordercolor(color) 0
./cbm610.h:145:#define _textcolor(color)       COLOR_WHITE
./cbm610.h:146:#define _bgcolor(color)         COLOR_BLACK
./cbm610.h:147:#define _bordercolor(color)     COLOR_BLACK
./cbm610.h:148:#define _cpeekcolor(color)      COLOR_WHITE
./cbm_filetype.h:94:unsigned char __fastcall__ _cbm_filetype (unsigned char c);
./pet.h:105:#define _textcolor(color)       COLOR_WHITE
./pet.h:106:#define _bgcolor(color)         COLOR_BLACK
./pet.h:107:#define _bordercolor(color)     COLOR_BLACK
./pet.h:108:#define _cpeekcolor(color)      COLOR_WHITE
./_gtia.h:108:#define _gtia_mkcolor(hue,lum) (((hue) << 4) | ((lum) << 1))
./_gtia.h:135:#define GTIA_COLOR_BLACK             _gtia_mkcolor(HUE_GREY,0)
./_gtia.h:136:#define GTIA_COLOR_WHITE             _gtia_mkcolor(HUE_GREY,7)
./_gtia.h:137:#define GTIA_COLOR_RED               _gtia_mkcolor(HUE_REDORANGE,1)
./_gtia.h:138:#define GTIA_COLOR_CYAN              _gtia_mkcolor(HUE_CYAN,3)
./_gtia.h:139:#define GTIA_COLOR_VIOLET            _gtia_mkcolor(HUE_PURPLE,4)
./_gtia.h:140:#define GTIA_COLOR_GREEN             _gtia_mkcolor(HUE_GREEN,2)
./_gtia.h:141:#define GTIA_COLOR_BLUE              _gtia_mkcolor(HUE_BLUE,2)
./_gtia.h:142:#define GTIA_COLOR_YELLOW            _gtia_mkcolor(HUE_YELLOW,7)
./_gtia.h:143:#define GTIA_COLOR_ORANGE            _gtia_mkcolor(HUE_ORANGE,5)
./_gtia.h:144:#define GTIA_COLOR_BROWN             _gtia_mkcolor(HUE_YELLOW,2)
./_gtia.h:145:#define GTIA_COLOR_LIGHTRED          _gtia_mkcolor(HUE_REDORANGE,6)
./_gtia.h:146:#define GTIA_COLOR_GRAY1             _gtia_mkcolor(HUE_GREY,2)
./_gtia.h:147:#define GTIA_COLOR_GRAY2             _gtia_mkcolor(HUE_GREY,3)
./_gtia.h:148:#define GTIA_COLOR_LIGHTGREEN        _gtia_mkcolor(HUE_GREEN,6)
./_gtia.h:149:#define GTIA_COLOR_LIGHTBLUE         _gtia_mkcolor(HUE_BLUE,6)
./_gtia.h:150:#define GTIA_COLOR_GRAY3             _gtia_mkcolor(HUE_GREY,5)
```

**Comment by pmor13** [comment]
As I see it:
```
#define X _x    /* wrong: _x may be user-defined */
#define _x X    /* wrong: _x may be user-defined */
void _x(void);  /* wrong: _x may be user-defined */
etc.
```
Solution: `__x` or `_X`.

**Comment by groessler** [comment]
So, for these "underscore" identifiers in `atari.h`, I got a ticking-off from Uz, because he said that if one includes `atari.h` he gets things from the "atari" name space. He didn't like the leading underscore.

Your list, if changed, might result in compatibility problems with old code. Do we want that?

**Comment by mrdudz** [comment]
> He didn't like the leading underscore.

It looks odd to me as well for that kind of functions... why arent they named eg atari_xy ?

> Your list, if changed, might result in compatibility problems with old code. Do we want that?

For the cases where this is a concern, the old names could perhaps be defined via macros if STD=CC65. Actually, exporting these functions using a different (new "save") name and defining the old names via macro for STD=CC65 is perhaps also another good way to fix those functions that are exported from a file that contains other standard functions

Lets first fix that list and move things around and identify what needs what treatment :) @kugelfuhr @spiro-trikaliotis :)

**Comment by groessler** [comment]
On 7/24/22 23:35, Bob Andrews wrote:
>
>     He didn't like the leading underscore.
>
> It looks odd to me as well for that kind of functions... why arent 
> they named eg atari_xy ?
>
Because most of them are named after their BASIC equvalent. And the 
BASIC version doesn't have a "atari_" prefix

**Comment by pmor13** [comment]
Re: `atari.h`: FYI: The 7.1.3 Reserved identifiers is only relevant for the C standard library. The `atari.h`, `_gtia.h`, etc. are not part of the C standard library. Or in cc65 are they?

**Comment by pmor13** [comment]
I think that the following macros are affected:
```
errno.h:
   53: #define errno   _errno
setjmp.h:
   46: #define setjmp  _setjmp         /* ISO insists on a macro */
signal.h:
   52: #define SIG_DFL         _sig_dfl
   53: #define SIG_IGN         _sig_ign
time.h:
   93: #  define CLOCKS_PER_SEC        _clocks_per_sec()
```

**Comment by mrdudz** [comment]
> Reserved identifiers is only relevant for the C standard library. The atari.h, _gtia.h, etc. are not part of the C standard library. Or in cc65 are they?

In practise IMHO treatening all header files and functions equally in that regard make sense - because many, if not most, programs will include some non standard header file or use some of the provided non standard library functions (that is one of the strengths of cc65 afterall). I'd use a similar reasoning as already used above - we'd avoid the "_symbol" form not because the standard dictates it, but because doing so will certainly solve more problems than if not.

**Comment by groessler** [comment]
On 7/25/22 00:00, pmor wrote:
>
> I think that the following macros are affected:
>
> |errno.h: 53: #define errno _errno setjmp.h: 46: #define setjmp 
> _setjmp /* ISO insists on a macro */ signal.h: 52: #define SIG_DFL 
> _sig_dfl 53: #define SIG_IGN _sig_ign time.h: 93: # define 
> CLOCKS_PER_SEC _clocks_per_sec() |

This list looks manageable. Plus `_poserror`, of course, or? :-)

I could provide a patch to 'double-underscore' them...

**Comment by mrdudz** [comment]
please see my updated post/list :)

**Comment by kugelfuhr** [comment]
> > Reserved identifiers is only relevant for the C standard library. The atari.h, _gtia.h, etc. are not part of the C standard library. Or in cc65 are they?
> 
> In practise IMHO treatening all header files and functions equally in that regard make sense - because many, if not most, programs will include some non standard header file or use some of the provided non standard library functions (that is one of the strengths of cc65 afterall). I'd use a similar reasoning as already used above - we'd avoid the "_symbol" form not because the standard dictates it, but because doing so will certainly solve more problems than if not. If we can relatively easily avoid such symbol name clashes, even if someone included pet.h or conio.,h or whatever, then why not do it.

In case of the non standard header files this doesn't make sense, because you will then need to rename __all__ identifiers including macro names. Not only the ones with underscores.

Let me reiterate what the problem is:
 * A conforming C compiler must be able to successfully translate every strictly conforming C program.
 * A strictly conforming program may use any identifier not reserved by the C standard.
 * So in turn the standard headers may not introduce such identifiers because otherwise a strictly conforming program may not compile successfully.

If you would extend this to the non standard headers, these headers wouldn't be able to use identifiers like `CH_ENTER`, `get_ostype` and so on, because any C program could use these identifiers and no collisions would be allowed. As a result you will need to move __all__ identifiers into the namespace reserved for the implementation, leading to `_CH_ENTER`, `_get_ostype` ....

I don't think that anybody really wants this. And in fact it isn't necessary, since a program including a header like "atari.h" is no longer strictly conforming and can no longer expect that the guarantees made by the C standard regarding allowed identifiers hold true.

**Comment by pmor13** [comment]
@kugelfuhr Per C `"atari.h"` is not a header. It is a source file (though, for the sake of grammar simplification it is `header-name`).

6.10.2 Source file inclusion:
> A preprocessing directive of the form
> ```
> # include "q-char-sequence" new-line
> ```
> causes the replacement of that directive by the entire contents of the **source file** identified by the specified sequence between the " delimiters.

**Comment by pmor13** [comment]
@mrdudz 
> we'd avoid the "_symbol" form not because the standard dictates it, but because doing so will certainly solve more problems than if not

The standard dictates it (7.1.3 Reserved identifiers) to avoid name collisions. As an implementor you are required to use only reserved identifiers because non-reserved identifiers shall be available to the user. In other words: the non-reserved identifiers are not your identifiers, but user's identifiers. If you reserve user's identifiers, then you violate the C standard. And, as a result, the user may not compile its (strictly conforming) program.

**Comment by groessler** [comment]
> I think that the following macros are affected:
> 
> ```
> errno.h:
>    53: #define errno   _errno
> setjmp.h:
>    46: #define setjmp  _setjmp         /* ISO insists on a macro */
> signal.h:
>    52: #define SIG_DFL         _sig_dfl
>    53: #define SIG_IGN         _sig_ign
> time.h:
>    93: #  define CLOCKS_PER_SEC        _clocks_per_sec()
> ```

Would a patch by me be accepted which converts those identifiers to use double _?
Then finally close this issue...

**Comment by mrdudz** [comment]
Please look again at the updated list - we should also take care of those functions that are exposed in the standard headers. Lets get rid of this issue once and for all now :)

**Comment by spiro-trikaliotis** [comment]
> Whatever you work out - please consider creating a test for it :)

Just a wild idea: Have a command-line option to the compiler that just compiles a file (for example, a header) and writes all identifiers that were defined into a file. This can be either one file, or separate files for every name space.

The test case could then check if only identifiers are defined that are not reserved.

**Comment by mrdudz** [comment]
> The test case could then check if only identifiers are defined that are not reserved.

For a similar case i have created this: https://github.com/cc65/cc65/issues/1796 - the backdraw of course is that such file would be required for every single symbol

**Comment by groessler** [comment]
On 7/25/22 21:28, Bob Andrews wrote:
>
> Please look again at the updated list - we should also take care of 
> those functions that are exposed in the standard headers. Lets get rid 
> of this issue once and for all now :)
>

Where is the updated list?


Maybe you could start by reverting the reversion of 1805?  :-)

**Comment by mrdudz** [comment]
> Where is the updated list? Maybe you could start by reverting the reversion of 1805?  :-)

*sigh* Right here, some posts above. (Participating via mail does not work, i wish i could disable this feature completely)

**Comment by groessler** [comment]
Ok. The list in the post with "I cant reopen that PR, so if you want to put some time into fixing this..." in the text?

**Comment by mrdudz** [comment]
yes, indeed

**Comment by mrdudz** [comment]
started fixing this and other similar problems in #1844 - please have a look and help finding other related problems

**Comment by mrdudz** [comment]
closed via #1844
